WebView/app/src/androidTest/java/com/android/samples/webviewdemo/MainActivityTest.kt

```diff
@@ -74,78 +82,82 @@ class MainActivityTest {
+            .check(
+                webMatches(
+                    script(
+                        "return jsObject && jsObject.postMessage != null;",
+                        castOrDie(Boolean::class.javaObjectType)
+                    ),
+                    `is`(true)
+                )
             )
-        )
     }
 
     @Test
-    fun valueInCallback_compareValueInput_returnsTrue(){
-        mainActivityRule.getActivity()
+    fun valueInCallback_compareValueInput_returnsTrue() = runBlocking() {
+        mainActivityRule.activity
         // Setup
         val jsObjName = "jsObject"
         val allowedOriginRules = setOf<String>("https://example.com")
-        val message = "hello"
+        val expectedMessage = "hello"
+        val callback = async { return@async "hello" }
         // Get a handler that can be used to post to the main thread
-        val mainHandler = Handler (Looper.getMainLooper());
-        val myRunnable = Runnable() {
-            val webView = WebView(context)
-            // Create JsObject
-            createJsObject(
-                webView,
-                jsObjName,
-                allowedOriginRules
-            ) { message -> //save message; call .set()
-            }
-            run() {
+        val mainHandler = Handler(Looper.getMainLooper());
+        mainHandler.post {
+            run {
+                val webView = WebView(context)
+                // Create JsObject
+                createJsObject(
+                    webView,
+                    jsObjName,
+                    allowedOriginRules
+                ) { message -> callback
+                }
                 //Inject JsObject into Html
-                webView.loadDataWithBaseURL("https://example.com","<html></html>",
-                    "text/html", "UTF-8","https://example.com")
+                webView.loadDataWithBaseURL(
+                    "https://example.com", "<html></html>",
+                    "text/html", null, null
+                )
                 //Call js code to invoke callback
-                webView.evaluateJavascript("${jsObjName}.postMessage(${message})", null)
+                webView.evaluateJavascript("${jsObjName}.postMessage(`hello`)", null)
             }
         }
-        mainHandler.post(myRunnable)
         // evaluate what comes out -> it should be hello
-        // *Note: //.get() is a place holder
-        assertEquals("hello"
-            , "//.get()"
-             )
+        assertEquals(expectedMessage, callback.await())
     }
 
     @Test
     // Checks that postMessage runs on the UI thread
-    fun checkingThreadCallbackRunsOn() {
-        mainActivityRule.getActivity()
+    fun checkingThreadCallbackRunsOn() = runBlocking {
+        mainActivityRule.activity
         // Setup
         val jsObjName = "jsObject"
         val allowedOriginRules = setOf<String>("https://example.com")
-        val message = "hello"
+        val callback = async { isUiThread() }
```

Comment:
Looking at this again: how does this control when the coroutine executes? I wonder if the coroutine is executing on its own before the callback happens, which might explain test failures (the code is running on a non-UI thread, so the assertion would fail).